lnrpc/routerrpc/router_server.go

```diff
@@ -401,11 +447,276 @@ func (s *Server) EstimateRouteFee(ctx context.Context,
 	return &RouteFeeResponse{
 		RoutingFeeMsat: int64(route.TotalFees()),
 		TimeLockDelay:  int64(route.TotalTimeLock),
+		FailureReason:  lnrpc.PaymentFailureReason_FAILURE_REASON_NONE,
 	}, nil
 }
 
-// SendToRouteV2 sends a payment through a predefined route. The response of this
-// call contains structured error information.
+// probeInvoice estimates fees along a route to a destination that is specified
+// in an invoice. The estimation duration is limited by a timeout. In case that
+// route hints are provided, this method applies a heuristic to identify LSPs
+// which might block probe payments. In that case, fees are manually calculated
+// and added to the probed fee estimation up until the LSP node. If the route
+// hints don't indicate an LSP, they are passed as arguments to the
+// SendPayment_V2 method, which enable it to send probe payments to the payment
+// request destination.
+func (s *Server) probeInvoice(ctx context.Context, paymentRequest string,
+	timeout int32) (*RouteFeeResponse, error) {
+
+	payReq, err := zpay32.Decode(
+		paymentRequest, s.cfg.RouterBackend.ActiveNetParams,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Generate random payment hash, so we can be sure that the target of
+	// the probe payment doesn't have the preimage to settle the htlc.
+	var paymentHash lntypes.Hash
+	_, err = crand.Read(paymentHash[:])
+	if err != nil {
+		return nil, fmt.Errorf("cannot generate random probe "+
+			"preimage: %w", err)
+	}
+
+	probeRequest := &SendPaymentRequest{
+		TimeoutSeconds:   timeout,
+		Dest:             payReq.Destination.SerializeCompressed(),
+		MaxParts:         1,
+		AllowSelfPayment: false,
+		AmtMsat:          int64(*payReq.MilliSat),
+		PaymentHash:      paymentHash[:],
+		FeeLimitSat:      100_000_000,
```

Comment:
nit: the query routes branch uses the same limit but written differently, a file-level constant or at least using the same constant would be nice